db/snapshot/backup_task.cc

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ *
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#include "utils/lister.hh"
+#include "utils/s3/client.hh"
+#include "replica/database.hh"
+#include "db/config.hh"
+#include "db/snapshot-ctl.hh"
+#include "db/snapshot/backup_task.hh"
+#include "schema/schema_fwd.hh"
+#include "sstables/sstables.hh"
+
+extern logging::logger snap_log;
+
+namespace db::snapshot {
+
+backup_task_impl::backup_task_impl(tasks::task_manager::module_ptr module, snapshot_ctl& ctl, shared_ptr<s3::client> client, sstring bucket, sstring ks, sstring snapshot_name) noexcept
+    : tasks::task_manager::task::impl(module, tasks::task_id::create_random_id(), 0, "node", ks, "", "", tasks::task_id::create_null_id())
+    , _snap_ctl(ctl)
+    , _client(std::move(client))
+    , _bucket(std::move(bucket))
+    , _ks(std::move(ks))
+    , _snapshot_name(std::move(snapshot_name))
+{}
+
+std::string backup_task_impl::type() const {
+    return "backup";
+}
+
+tasks::is_internal backup_task_impl::is_internal() const noexcept {
+    return tasks::is_internal::no;
+}
+
+void backup_task_impl::run_in_thread(sstring data_dir) {
+    std::exception_ptr ex;
+    //
+    // The keyspace data directories and their snapshots are arranged as follows:
+    //
+    //  <data dir>
+    //  |- <keyspace name1>
+    //  |  |- <column family name1>
+    //  |     |- snapshots
+    //  |        |- <snapshot name1>
+    //  |          |- <snapshot file1>
+    //  |          |- <snapshot file2>
+    //  |          |- ...
+    //  |        |- <snapshot name2>
+    //  |        |- ...
+    //  |  |- <column family name2>
+    //  |  |- ...
+    //  |- <keyspace name2>
+    //  |- ...
+    //
+    auto ks_dir = fs::path(data_dir) / _ks;
+    auto ks_dir_lister = directory_lister(ks_dir, lister::dir_entry_types::of<directory_entry_type::directory>());
+    auto close_ks_dir_lister = deferred_close(ks_dir_lister);
+
+    snap_log.trace("Scan keyspace dir {}", ks_dir.native());
+    while (auto table_ent = ks_dir_lister.get().get()) {
+        auto snapshot_dir = ks_dir / table_ent->name / sstables::snapshots_dir / _snapshot_name;
+        auto has_snapshot = file_exists(snapshot_dir.native()).get();
+
+        snap_log.trace("Check table snapshot dir {} (exists={})", snapshot_dir.native(), has_snapshot);
+        if (!has_snapshot) {
+            continue;
+        }
+
+        auto cf_name_and_uuid = replica::parse_table_directory_name(table_ent->name);
+        auto snapshot_dir_lister = directory_lister(snapshot_dir, lister::dir_entry_types::of<directory_entry_type::regular>());
+        auto close_snapshot_dir_lister = deferred_close(snapshot_dir_lister);
+
+        while (auto component_ent = snapshot_dir_lister.get().get()) {
+            auto gh = _uploads.hold();
```

Comment:
Sorry, but I still dislike synchronization via the _uploads gate.
I think it would be better if this function would have its own internal gate and it should await this gate before returning.

An external gate makes sense if it could be closed asynchronously, e.g. on shutdown.
That said, do we need an abort source to abort this process on global shutdown, or would that be done via the task abort method, which should trigger when aborting the whole task module?

Reply:
> I think it would be better if this function would have its own internal gate and it should await this gate before returning.

The run_in_thread() one?

> That said, do we need an abort source to abort this process on global shutdown, or would that be done via the task abort method, which should trigger when aborting the whole task module?

It should be done via task abort method. It's not in this PR yes, but it should also be possible to abort backup task even _not_ on shutdown, but just via an API call. For that we'll need a way to abort s3::upload_file() too, but it currently doesn't allow that

Reply:
> For that we'll need a way to abort s3::upload_file() too, but it currently doesn't allow that

Yes, but even aborting in-between uploads is better than nothing

Reply:
> > I think it would be better if this function would have its own internal gate and it should await this gate before returning.
> 
> The run_in_thread() one?

Yes. (or just run(), if it creates the thread for itself)

Currently it spawns background fibers inside the `_uploads` gate and the caller is responsible for closing the gate.  It feels like "programming by side-effects" to me, and ti would be simpler if the run() function would just return a future that would wait for all the background fibers it launched, i.e. call `local_gate.close().get()` in the seastar thread lambda.

> 
> > That said, do we need an abort source to abort this process on global shutdown, or would that be done via the task abort method, which should trigger when aborting the whole task module?
> 
> It should be done via task abort method. It's not in this PR yes, but it should also be possible to abort backup task even _not_ on shutdown, but just via an API call. For that we'll need a way to abort s3::upload_file() too, but it currently doesn't allow that

Reply:
> Yes, but even aborting in-between uploads is better than nothing

Sure. It's in nearest plans, the task should be abortable

Reply:
> Yes. (or just run(), if it creates the thread for itself)

If the gate is internal to run_in_thread() then it should also catch its own internal errors not to return with non-closed gate. This increases the indentation I'm trying to avoid.

Making gate inside run() and passing it by reference into run_in_thread() sounds like
> ... it spawns background fibers inside the _uploads gate and the caller is responsible for closing the gate.

Reply:
Well, we can wait for gate in defer :woozy_face: @bhalevy , please check two top-most commits from [this branch](https://github.com/xemul/scylla/commits/br-bacup-to-object-store-api-call-A/), does it look like what you have in mind?

Reply:
yes, thanks. much better. this way the run function is much more self-contained.